server/src/persistence/file.rs

```diff
@@ -0,0 +1,297 @@
+
+use crate::types::EdgeToken;
+use crate::{
+    error::EdgeError,
+    types::{EdgeResult, TokenRefresh},
+};
+
+use super::EdgePersistence;
+
+pub struct FilePersister {
+    pub storage_path: PathBuf,
+}
+
+impl TryFrom<&str> for FilePersister {
+    type Error = EdgeError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        PathBuf::from_str(value)
+            .map(|path| Self { storage_path: path })
+            .map_err(|_e| {
+                EdgeError::PersistenceError(format!("Could not build a path from {value}"))
+            })
+    }
+}
+
+impl FilePersister {
+    pub fn token_path(&self) -> PathBuf {
+        let mut token_path = self.storage_path.clone();
+        token_path.push("unleash_tokens.json");
+        token_path
+    }
+
+    pub fn features_path(&self) -> PathBuf {
+        let mut features_path = self.storage_path.clone();
+        features_path.push("unleash_features.json");
+        features_path
+    }
+
+    pub fn refresh_target_path(&self) -> PathBuf {
+        let mut refresh_target_path = self.storage_path.clone();
+        refresh_target_path.push("unleash_refresh_targets.json");
+        refresh_target_path
+    }
+
+    pub fn new(storage_path: &Path) -> Self {
+        FilePersister {
+            storage_path: storage_path.to_path_buf(),
+        }
+    }
+}
+
+#[async_trait]
+impl EdgePersistence for FilePersister {
+    async fn load_tokens(&self) -> EdgeResult<Vec<EdgeToken>> {
+        let mut file = tokio::fs::File::open(self.token_path())
+            .await
+            .map_err(|_| {
+                EdgeError::PersistenceError(
+                    "Cannot load tokens from backup, opening backup file failed".to_string(),
+                )
+            })?;
+
+        let mut contents = vec![];
+
+        file.read_to_end(&mut contents).await.map_err(|_| {
+            EdgeError::PersistenceError(
+                "Cannot load tokens from backup, reading backup file failed".to_string(),
+            )
+        })?;
+        serde_json::from_slice(&contents).map_err(|_| {
+            EdgeError::PersistenceError(
+                "Cannot load tokens from backup, parsing backup file failed".to_string(),
+            )
+        })
+    }
+
+    async fn save_tokens(&self, tokens: Vec<EdgeToken>) -> EdgeResult<()> {
+        let mut file = tokio::fs::File::create(self.token_path())
+            .await
+            .map_err(|_| {
+                EdgeError::PersistenceError(
+                    "Cannot write tokens to backup. Opening backup file for writing failed"
+                        .to_string(),
+                )
+            })?;
+        file.write_all(
+            &serde_json::to_vec(&tokens)
+                .map_err(|_| EdgeError::PersistenceError("".to_string()))?,
+        )
+        .await
+        .map_err(|_| EdgeError::PersistenceError("Could not serialize tokens to disc".to_string()))
+        .map(|_| ())
+    }
```

Comment:
When I see writing to file and perhaps son errors along the way, I tend to think that writing to a temporary file and then swapping would be a good idea to protect from having an unrecoverable backup file

Reply:
The actual write is a single pass but yeah, writing to a file can fail half way through. It'd be a nice improvement but this beast is big enough right now